File: Multidimensional/multidimensional-v1.py
import matplotlib.pyplot as plt
import numpy as np

sigma_acel = 0.2                    #Desvio padrao da aceleracao
sigma_medicao = 3                   #Desvio padrao da medicao
tempo = 1                           #Variacao de tempo

x = np.zeros((6, 1))                #Inicia a matriz 6x1 de estimativa com zeros

F = np.zeros((6, 6))                #Inicia a matriz F 6x6 e a preenche com os respectivos valores
for i in range(0, 6):
    F[i, i] = 1                     #Preenche a diagonal principal
F[0, 1] = F[1, 2] = F[3, 4] = F[4, 5] = tempo
F[0, 2] = F[3, 5] = 0.5 * (tempo ** 2)

Q = np.zeros((6, 6))                #Inicia a matriz Q 6x6 e a preenche com os respectivos valores
Q[0, 0] = Q[3, 3] = (tempo**4)/4
Q[0, 1] = Q[1, 0] = Q[3, 4] = Q[4, 3] = (tempo**3)/2
Q[0, 2] = Q[2, 0] = Q[3, 5] = Q[5, 3] = (tempo**2)/2
Q[1, 1] = Q[4, 4] = tempo**2
Q[1, 2] = Q[2, 1] = Q[4, 5] = tempo
Q[2, 2] = Q[5, 5] = 1
Q = Q * (sigma_acel**2)

R = np.zeros((2, 2))                #Inicia a matriz R 2x2 e a preenche com os respectivos valores
R[0, 0] = R[1, 1] = sigma_medicao**2

P = np.zeros((6, 6))                #Inicia a matriz P 6x6 e a preenche com valores arbitrarios
for i in range(0, 6):
    P[i, i] = 500                   #Preenche a diagonal principal

H = np.zeros((2, 6))                #Inicia a matriz H 2x6 e a preenche com valores arbitrarios
H[0, 0] = H[1, 3] = 1

I = np.zeros((6, 6))                #Inicia a matriz I identidade 6x6 e a preenche com os respectivos valores
for i in range(0, 6):
    I[i, i] = 1

Ft = np.transpose(F)                #Inicia a matriz transposta de F
Ht = np.transpose(H)                #Inicia a matriz transposta de H

File: Multidimensional/multidimensional-v1.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% DADOS INICIAIS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sigma_acel = 0.2*0.2;                                   %Constante
sigma_medicao = 3*3;                                    %Constante

x = zeros(6, 1);                                        %Inicializacao da matriz 6x1 de estimativa

Q = [0.25 0.5 0.5 0 0 0 ; 0.5 1 1 0 0 0 ; 0.5 1 1 0 0 0 ; 0 0 0 0.25 0.5 0.5 ; 0 0 0 0.5 1 1 ; 0 0 0 0.5 1 1]*(sigma_acel);

P = [500 0 0 0 0 0 ; 0 500 0 0 0 0 ; 0 0 500 0 0 0 ; 0 0 0 500 0 0 ; 0 0 0 0 500 0 ; 0 0 0 0 0 500];

F = [1 1 0.5 0 0 0 ; 0 1 1 0 0 0 ; 0 0 1 0 0 0 ; 0 0 0 1 1 0.5 ; 0 0 0 0 1 1 ; 0 0 0 0 0 1];

H = [1 0 0 0 0 0 ; 0 0 0 1 0 0 ];

R = [sigma_medicao 0 ; 0 sigma_medicao];

I = eye(6);                                             %Matriz identidade 6x6

Ft = transpose(F);
Ht = transpose(H);

z = zeros(2, 1);                                        %Inicializacao da matriz 2x1 de medicao

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

load medidas_v1.mat
N = length(medx);

estimados = zeros(6, N);                               %Inicializacao da matriz de estimativa, com x dados em i tempos
medidos = zeros(2, N);                                 %Inicializacao da matriz auxiliar de medicao, com x dados em i tempos

%Laco de repeticao para medicao e estimacao de i tempos
for i = 1:N

    x = F*x;                                            %Predicao da estimativa 
    P = F*P*Ft + Q;                                     %Predicao da covariancia

    %Entrada da matriz z pelo usuario                   
    z(1) = medx(i);
    z(2) = medy(i);

    K = P*Ht*(H*P*Ht + R)^(-1);                         %Predicao do ganho de Kalman
    Kt = transpose(K);                                  %Atualizacao da matriz transposta do ganho de Kalman                                  

    x = x + K*(z - H*x);                                %Atualizacao da matriz de estimativa
    P = (I - K*H)*P*(transpose(I - K*H)) + K*R*Kt;      %Atualizacao da matriz de covariancia

    estimados(:,i) = x;                                 %Matriz de estimados recebe x, no tempo i
    medidos(:,i) = z;                                   %Matriz de medidos recebe z, no tempo i                               

end

estimadoX = estimados(1,:);                              %Armazena as estimativas da posicao X
estimadoY = estimados(4,:);                              %Armazena as estimativas da posicao Y 
medidoX = medidos(1,:);
medidoY = medidos(2,:);

%Plot do grafico de estimativa da posicao XY
figure
plot(estimadoX, estimadoY, 'b')
hold on
scatter(medidoX,medidoY,'r')

xlabel('Posicao X (m)');
ylabel('Posicao Y (m)');
title('Estimativa vs Medicao da posicao XY');
legend('estimado','medido');
grid on;
